test/hotspot/jtreg/gc/TestNativeReferenceGet.java

```diff
@@ -0,0 +1,175 @@
+    // old generation.
+    private static void setup() {
+        references = new ArrayList<Ref>(NUM_REFS);
+        referents = new ArrayList<TestObject>(NUM_REFS);
+
+        for (int i = 0; i < NUM_REFS; ++i) {
+            TestObject obj = new TestObject(i);
+            referents.add(obj);
+            references.add(new Ref(obj));
+        }
+
+        gcUntilOld(references);
+        gcUntilOld(referents);
+        for (int i = 0; i < NUM_REFS; ++i) {
+            gcUntilOld(references.get(i));
+            gcUntilOld(referents.get(i));
+        }
+    }
+
+    // Discard all the strong references.
+    private static void clearReferents() {
+        // Not using List.clear() because it doesn't document null'ing elements.
+        for (int i = 0; i < NUM_REFS; ++i) {
+            referents.set(i, null);
+        }
+    }
+
+    // Create new strong references from the weak references, by using the
+    // native method implementation of Reference.get() and recording the value
+    // in references.
+    private static void strengthenReferents() {
+        for (int i = 0; i < NUM_REFS; ++i) {
+            referents.set(i, references.get(i).get());
+        }
+    }
+
+    private static final long TIMEOUT = 10 * 1000; // 10 seconds, in millis.
+
+    private static void check() {
+        // None of the references should have been cleared and enqueued,
+        // because we have strong references to all the referents.
+        try {
+            if (queue.remove(TIMEOUT) != null) {
+                throw new RuntimeException("Reference enqueued");
+            }
+        } catch (InterruptedException e) {
+            throw new RuntimeException("Test interrupted");
+        }
+        // Check details of expected state.
+        for (int i = 0; i < NUM_REFS; ++i) {
+            Ref reference = (Ref) references.get(i);
+            TestObject referent = reference.get();
+            if (referent == null) {
+                throw new RuntimeException("Referent not strengthened");
+            } else if (referent != referents.get(i)) {
+                throw new RuntimeException(
+                    "Reference referent differs from saved referent: " + i);
+            } else if (referent.value != i) {
+                throw new RuntimeException(
+                    "Referent " + i + " value: " + referent.value);
+            }
+        }
+    }
+
+    private static void testConcurrent() {
+        System.out.println("Testing concurrent GC");
+        try {
+            WB.concurrentGCAcquireControl();
+            clearReferents();
+            WB.concurrentGCRunTo(WB.BEFORE_MARKING_COMPLETED);
+            strengthenReferents();
+            WB.concurrentGCRunToIdle();
+            check();
+        } finally {
+            WB.concurrentGCReleaseControl();
+        }
+    }
+
+    private static void testNonconcurrent() {
+        System.out.println("Testing nonconcurrent GC");
+        clearReferents();
+        strengthenReferents();
```

Comment:
Might the GC clear refs between `clearReferents()` and `strengthenReferents()`?

Reply:
Yeah, an ill-timed GC between those operations would result in test failure.
I've added a GC immediately before those operations to make that pretty unlikely.